### PR TITLE
Add upper energy cut-off to ThomasFermi model

### LINKED ISSUE #1967: Implement energy cut-off for Thomas-Fermi model
This is a follow-up issue to #1444.

In order to calculate a local "temperature" only from the *bulk* / *thermal* electrons in the plasma an energy cut-off has to be implemented for the `ThomasFermi` impact ionization model. 

- [x] implement the cut-off
- [x] add user documentation

### PR DESCRIPTION
**Pull request by n01r**
Separate bulk from prompt electrons by defining an energy cut-off.
Above this threshold value electrons are classified as "prompt" electrons
and are therefore neglected in the calculation of electron temperature.

The actual cut-off value is highly debated, see [Kemp, Sentoku](http://link.aps.org/doi/10.1103/PhysRevLett.116.159501).

This is one of 4 planned improvements of the `ThomasFermi` implementation.

A test with a directed stream of electrons has shown that if the electrons have energies higher than the cut-off they will not be considered for the "temperature" calculation.
The test features a 2-dimensional 256 x 256 grid with period boundaries in every direction and about 6 copper ions distributed in every cell.
The ions are preionized once and electrons are derived from them.
There are three test setups:
1. electron temperature of 0 eV
2. electron temperature of 100 eV
3. electron temperature of 0 eV and drift of 200 keV
An energy cut-off of 50 keV has been configured.

The results show that in case 1 and 2 the predicted charge state matches the one reached in the simulation.
The TF model predicts an initial charge state for copper at density 8.92 g/cm³ and T_e = 0 eV of 4+.
Case 3 performs exactly as case 1 until the fast electrons distribute some of their energy to the cold electrons and TF ionization continues.

~~- [ ] Perform a more advanced test with disabled current deposition and a velocity gradient that depends on the cell index perpendicular to the stream direction   
Thereby redistribution of electron energies is avoided and we should see the cut-off spatially in a jump of ion charge state / electron density~~
**update 2017-11-07**   
A simple test has been done and it shows that the implementation works *BUT* a more detailed test will be relocated to the issue #1967.

Expectation:
 
04_high_energy_cutoff_gradient_test_expectation

- [x] Add documentation to the `/docs` in the TF section

cc'ing @HighIander 

This PR closes #1967 .

### CODE REVIEW

File: include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        /* read existing attributes */
+        const float_X weighting = particle[weighting_];
+        const float3_X mom = particle[momentum_];
+        const float_X mass = attribute::getMass( weighting, particle );
+
+        constexpr float_X INV_CELL_VOLUME = float_X(1.0) / CELL_VOLUME;
+
+        /* value for energy cut-off */
+        constexpr float_X cutoff_energy = particles::ionization::thomasFermi::TF_CUTOFF_ENERGY;
```
Comment (psychocoderHPC) [inline include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp]: The functor `EnergyDensityCutoff` is part of the gerneric namespace `particleToGrid` but depends on values of an special ionization model *ThomasFermi*. 
This dependency is not mentioned in the documentation in the `*.def` file.
We need to pass the value of `particles::ionization::thomasFermi::TF_CUTOFF_ENERGY` somehow in the template interface or via constructor if the functor concept used following http://picongpu.readthedocs.io/en/latest/usage/particles.html?highlight=filter#manipulation-functors.

File: include/picongpu/simulation_defines/param/ionizer.param [hunk 2]
```diff
@@ -423,6 +423,20 @@ namespace thomasFermi
...
+     * of electrons. For the Thomas-Fermi ionization model we only want the
+     * thermalized "bulk" electrons. Including the high-energy "prompt"
+     * is physically questionable since they do not have a large cross section
+     * for collisional ionization.
+     *
+     * unit: keV
+     */
+    constexpr float_X CUTOFF_ENERGY_KEV = 50;
+    /** cutoff energy for electron "temperature" calculation in SI units*/
+    constexpr float_X TF_CUTOFF_ENERGY = CUTOFF_ENERGY_KEV * UNITCONV_keV_to_Joule;
```
Comment (ax3l) [inline include/picongpu/simulation_defines/param/ionizer.param]: naming of those two vars: I think here is no `TF_` prefix needed (namespace is thomasFermi)

File: include/picongpu/simulation_defines/param/ionizer.param [hunk 3]
```diff
@@ -423,6 +423,20 @@ namespace thomasFermi
     constexpr float_X TFC1    = -0.366667;
     constexpr float_X TFC2    =  0.983333;
 
+    /** cutoff energy for electron "temperature" calculation
+     *
+     * In laser produced plasmas we can have different, well-separable groups
+     * of electrons. For the Thomas-Fermi ionization model we only want the
+     * thermalized "bulk" electrons. Including the high-energy "prompt"
```
Comment (ax3l) [inline include/picongpu/simulation_defines/param/ionizer.param]: missing subject: add "[...] prompt *electrons*"

Comment (ax3l) [reply include/picongpu/simulation_defines/param/ionizer.param]: this new cut-off and the upcoming one should in included in the model description of the `docs/`.

You can e.g. just take a variant of your good PR description.

File: include/picongpu/particles/ionization/byCollision/ThomasFermi/ThomasFermi_Impl.hpp [hunk 4]
```diff
@@ -129,12 +129,12 @@ namespace ionization
...
             using DensitySolver = typename particleToGrid::CreateDensityOperation<T_SrcSpecies>::type::Solver;
 
-            /** Solver for energy density of the electron species
+            /** Solver for energy density of the electron species with energy cutoff
              *
              *  @todo Include all electron species with a ForEach<VectorallSpecies,...>
              * instead of just the destination species
              */
-            using EnergyDensitySolver = typename particleToGrid::CreateEnergyDensityOperation<T_DestSpecies>::type::Solver;
+            using EnergyDensitySolver = typename particleToGrid::CreateEnergyDensityCutoffOperation<T_DestSpecies>::type::Solver;
```
Comment (ax3l) [inline include/picongpu/particles/ionization/byCollision/ThomasFermi/ThomasFermi_Impl.hpp]: his line is really long, feel free to write
```C++
using EnergyDensitySolver = typename particleToGrid::CreateEnergyDensityCutoffOperation<
    T_DestSpecies
>::type::Solver;
```

File: include/picongpu/particles/ionization/byCollision/ThomasFermi/ThomasFermi_Impl.hpp [hunk 5]
```diff
@@ -129,12 +129,12 @@ namespace ionization
              */
             using DensitySolver = typename particleToGrid::CreateDensityOperation<T_SrcSpecies>::type::Solver;
 
-            /** Solver for energy density of the electron species
+            /** Solver for energy density of the electron species with energy cutoff
```
Comment (ax3l) [inline include/picongpu/particles/ionization/byCollision/ThomasFermi/ThomasFermi_Impl.hpp]: with *maximum* energy cutoff

File: include/picongpu/simulation_defines/param/ionizer.param [hunk 6]
```diff
@@ -423,6 +423,20 @@ namespace thomasFermi
...
+     * is physically questionable since they do not have a large cross section
+     * for collisional ionization.
+     *
+     * unit: keV
+     */
+    constexpr float_X CUTOFF_ENERGY_KEV = 50;
```
Comment (ax3l) [inline include/picongpu/simulation_defines/param/ionizer.param]: float: `50.0`

**Comment by n01r** [comment]
@ax3l , @psychocoderHPC thanks for the review! I'll see to the implementation of your improvements!

I will put up the other two cutoffs soon, as well. The test I thought off would be able to test all three cutoffs separately or even together but that thing might take me some more time. I'd rather get the first steps done first.

cutoff_test_setup

**Comment by ax3l** [comment]
cool, if that would be actual physics it would be a great Biermann battery vs. shear surface instability setup xD

**Comment by n01r** [comment]
Well, let's hope that at least the implementation of a position-dependent drift energy could be useful later. :sweat_smile:

File: include/picongpu/particles/ionization/byCollision/ThomasFermi/ThomasFermi_Impl.hpp [hunk 7]
```diff
@@ -129,12 +136,15 @@ namespace ionization
...
-            /** Solver for energy density of the electron species with energy cutoff
+            /** Solver for energy density of the electron species with maximum energy cutoff
              *
              *  @todo Include all electron species with a ForEach<VectorallSpecies,...>
              * instead of just the destination species
              */
-            using EnergyDensitySolver = typename particleToGrid::CreateEnergyDensityCutoffOperation<T_DestSpecies>::type::Solver;
+            using EnergyDensitySolver = typename particleToGrid::CreateEnergyDensityCutoffOperation<
+                T_DestSpecies,
+                CutoffMaxEnergy
```
Comment (n01r) [inline include/picongpu/particles/ionization/byCollision/ThomasFermi/ThomasFermi_Impl.hpp]: Now the cutoff is more generic and the connection to `ThomasFermi` happens only here.
I still feel that the place/way I chose is maybe not too elegant.

File: include/picongpu/simulation_defines/param/ionizer.param (hunk 3 above)
Comment (n01r) [reply include/picongpu/simulation_defines/param/ionizer.param]: I added a description in `/docs` and also the missing subject has been added - just on a new line which is why the comment won't vanish. :smile:

**Comment by n01r** [comment]
@psychocoderHPC can you check if the [current implementation](https://github.com/ComputationalRadiationPhysics/picongpu/pull/2330/files#diff-52b1257e87c44a743db37fbd2f6493a7R146) is more to your liking?

@ax3l is [this](https://github.com/ComputationalRadiationPhysics/picongpu/pull/2330/files#diff-8b6cd506315fdce83575ca4cd661f51bR40) how you intended?

File: include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+        constexpr float_X INV_CELL_VOLUME = float_X(1.0) / CELL_VOLUME;
+
+        /* value for energy cut-off */
+        float_X const cutoff_max_energy = ParamClass::cutoff_max_energy;
+        float_X const cutoff = cutoff_max_energy / UNIT_ENERGY * weighting;
+
+        float_X const kinEnergy = KinEnergy<>()( mom, mass );
+
+        if (kinEnergy < cutoff)
```
Comment (psychocoderHPC) [inline include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp]: - please remove the space after the `if`
- add space after `(` and before `)`

File: include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+    template< class T_ParamClass >
+    template< class T_Particle >
+    DINLINE float_X
+    EnergyDensityCutoff< T_ParamClass >::operator()( T_Particle& particle ) const
+    {
+        using ParamClass =  T_ParamClass;
+
+        /* read existing attributes */
+        const float_X weighting = particle[weighting_];
```
Comment (psychocoderHPC) [inline include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp]: - please add spaces `particle[ weighting_ ]`
- move const to the right of the type `float_X const weighting = ...` (please check also the two lines after this line)

File: include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        /* value for energy cut-off */
+        float_X const cutoff_max_energy = ParamClass::cutoff_max_energy;
+        float_X const cutoff = cutoff_max_energy / UNIT_ENERGY * weighting;
+
+        float_X const kinEnergy = KinEnergy<>()( mom, mass );
+
+        if (kinEnergy < cutoff)
+            return kinEnergy * INV_CELL_VOLUME;
+
+        return float_X(0.);
```
Comment (psychocoderHPC) [inline include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp]: spaces missing after and before `0.`

File: include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+        /* read existing attributes */
+        const float_X weighting = particle[weighting_];
+        const float3_X mom = particle[momentum_];
+        const float_X mass = attribute::getMass( weighting, particle );
+
+        constexpr float_X INV_CELL_VOLUME = float_X(1.0) / CELL_VOLUME;
+
+        /* value for energy cut-off */
+        float_X const cutoff_max_energy = ParamClass::cutoff_max_energy;
```
Comment (psychocoderHPC) [inline include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp]: please use camel case names https://github.com/ComputationalRadiationPhysics/contributing/blob/master/codingGuideLines/cpp.md#3-global-rules

File: include/picongpu/particles/ionization/byCollision/ThomasFermi/ThomasFermi_Impl.hpp [hunk 12]
```diff
@@ -98,6 +98,13 @@ namespace ionization
 
         BlockArea BlockDescription;
 
+        /* parameter class containing the energy cutoff parameter for electron temperature calculation */
+        struct CutoffMaxEnergy
+        {
+            static constexpr float_X cutoff_max_energy =
```
Comment (psychocoderHPC) [inline include/picongpu/particles/ionization/byCollision/ThomasFermi/ThomasFermi_Impl.hpp]: Please use camel case names `cutoffMaxEnergy`

File: include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.def [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+namespace picongpu
+{
+namespace particleToGrid
+{
+namespace derivedAttributes
+{
+    /* Inherits from the `EnergyDensity` derived attribute
+     *
+     * \tparam T_ParamClass parameter class containing the maximum energy cutoff
```
Comment (psychocoderHPC) [inline include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.def]: please use `@` for doxygen commands

File: include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.def [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+#include "picongpu/particles/particleToGrid/derivedAttributes/EnergyDensity.def"
+
+
+namespace picongpu
+{
+namespace particleToGrid
+{
+namespace derivedAttributes
+{
+    /* Inherits from the `EnergyDensity` derived attribute
```
Comment (psychocoderHPC) [inline include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.def]: Must begin with `/**`

File: include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    template< class T_Particle >
+    DINLINE float_X
+    EnergyDensityCutoff< T_ParamClass >::operator()( T_Particle& particle ) const
+    {
+        using ParamClass =  T_ParamClass;
+
+        /* read existing attributes */
+        const float_X weighting = particle[weighting_];
+        const float3_X mom = particle[momentum_];
+        const float_X mass = attribute::getMass( weighting, particle );
```
Comment (psychocoderHPC) [inline include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp]: must be 
```C++
float_X const mass = attribute::getMass( 
    weighting, 
    particle 
);
```

File: include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+        constexpr float_X INV_CELL_VOLUME = float_X(1.0) / CELL_VOLUME;
+
+        /* value for energy cut-off */
+        float_X const cutoff_max_energy = ParamClass::cutoff_max_energy;
+        float_X const cutoff = cutoff_max_energy / UNIT_ENERGY * weighting;
+
+        float_X const kinEnergy = KinEnergy<>()( mom, mass );
```
Comment (psychocoderHPC) [inline include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp]:
```C++
float_X const kinEnergy = KinEnergy<>()(
    mom, 
    mass 
);
```

File: include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.def [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+         * \tparam T_Particle particle in the frame
+         * \param particle particle in the frame
+         *
+         * \return new attribute for the particle (type \see T_AttributeType)
+         */
+        template< class T_Particle >
+        DINLINE float_X
+        operator()( T_Particle& particle ) const;
+    };
+} /* namespace derivedAttributes */
```
Comment (psychocoderHPC) [inline include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.def]: please use `// namespace derivedAttributes`

File: include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.def [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        /** Calculate a new attribute  per particle
+         *
+         * Returns a new (on-the-fly calculated) attribute of a particle
+         * that can then be mapped to the cells the particle contributes to.
+         * This method is called on a per-thread basis (each thread of a block
+         * handles a particle of a frame).
+         *
+         * \tparam T_Particle particle in the frame
```
Comment (psychocoderHPC) [inline include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.def]: `@tparam`

File: include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp (hunk 16 above)
Comment (n01r) [reply include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp]: Shouldn't it be 
```c++
float_X const kinEnergy = KinEnergy< >( )(
    mom, 
    mass 
);
```
?

File: include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+        /* value for energy cut-off */
+        float_X const cutoffMaxEnergy = ParamClass::cutoffMaxEnergy;
+        float_X const cutoff = cutoffMaxEnergy / UNIT_ENERGY * weighting;
+
+        float_X const kinEnergy = KinEnergy< >( )(
+            mom,
+            mass
+        );
+
+        if( kinEnergy < cutoff )
```
Comment (psychocoderHPC) [inline include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.hpp]: Could you please change this to the following code. The code is equivalent to yours but an easy pattern which can be better understand by the compiler:

```C++
float_X result( 0. );
if( kinEnergy < cutoff )
    result = kinEnergy * INV_CELL_VOLUME;

 return result;
```

File: docs/source/models/collisional_ionization.rst [hunk 20]
```diff
@@ -12,16 +12,29 @@ Implemented LTE Model: Thomas-Fermi Ionization according to [More1985]_
 
 Get started here https://github.com/ComputationalRadiationPhysics/picongpu/wiki/Ionization-in-PIConGPU
 
+The implementation of the Thomas-Fermi model takes the following input quantities.
+
+- Ion proton number :math:`Z`
```
Comment (ax3l) [inline docs/source/models/collisional_ionization.rst]: lower case `i` as below

File: include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.def [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+{
+namespace particleToGrid
+{
+namespace derivedAttributes
+{
+    /** Inherits from the `EnergyDensity` derived attribute
+     *
+     * @tparam T_ParamClass parameter class containing the maximum energy cutoff
+     *
+     * Note: `T_ParamClass` requires the member `float_X constexpr cutoffMaxEnergy`.
```
Comment (ax3l) [inline include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.def]: `constexpr` to the left please :) (yes, it's an exception)

Comment (n01r) [reply include/picongpu/particles/particleToGrid/derivedAttributes/EnergyDensityCutoff.def]: whoops! right, didn't make that mistake in the code but in the comment ^^

**Pull request merged**